src/meta_protocol_proxy/upstream_handler_impl.cc

```diff
@@ -0,0 +1,162 @@
+#include "src/meta_protocol_proxy/upstream_handler_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace MetaProtocolProxy {
+
+UpstreamHandlerImpl::~UpstreamHandlerImpl() {
+  ENVOY_LOG(trace, "********** UpstreamHandlerImpl destructed ***********");
+  if (upstream_handle_ != nullptr) {
+    upstream_handle_->cancel(ConnectionPool::CancelPolicy::CloseExcess);
+    upstream_handle_ = nullptr;
+    ENVOY_LOG(debug, "UpstreamHandlerImpl: reset connection pool handler");
+  }
+}
+
+void UpstreamHandlerImpl::onClose() {
+  ENVOY_LOG(debug, "UpstreamHandlerImpl[{}] onClose", key_);
+  delete_callback_(key_);
+
+  if (upstream_handle_) {
+    ASSERT(!conn_data_);
+    upstream_handle_->cancel(Tcp::ConnectionPool::CancelPolicy::Default);
+    upstream_handle_ = nullptr;
+    ENVOY_LOG(debug, "UpstreamHandlerImpl::onClose reset connection pool handler");
+  }
+}
+
+int UpstreamHandlerImpl::start(Upstream::TcpPoolData& pool_data) {
+  Tcp::ConnectionPool::Cancellable* handle = pool_data.newConnection(*this);
+  if (handle) {
+    ASSERT(upstream_handle_ == nullptr);
+    upstream_handle_ = handle;
+  }
+  return 0;
+}
+
+void UpstreamHandlerImpl::onPoolFailure(ConnectionPool::PoolFailureReason reason,
+                                        absl::string_view transport_failure_reason,
+                                        Upstream::HostDescriptionConstSharedPtr host) {
+  ENVOY_LOG(error, "UpstreamHandlerImpl onPoolFailure [{}]:{} {} {}", key_,
+            host->address()->asString(), static_cast<int>(reason), transport_failure_reason);
+  upstream_handle_ = nullptr;
+  delete_callback_(key_);
+}
+
+void UpstreamHandlerImpl::onPoolReady(Tcp::ConnectionPool::ConnectionDataPtr&& conn_data,
+                                      Upstream::HostDescriptionConstSharedPtr host) {
+  ENVOY_LOG(debug, "UpstreamHandlerImpl onPoolReady [{}]:{}", key_, host->address()->asString());
+  upstream_host_ = host;
+  conn_data_ = std::move(conn_data);
+  upstream_handle_ = nullptr;
+  conn_data_->addUpstreamCallbacks(*this);
+  pool_ready_ = true;
+
+  ENVOY_CONN_LOG(debug, "UpstreamHandlerImpl[{}]: request_caches_ size:{}",
+                 conn_data_->connection(), key_, request_caches_.size());
+  for (auto& request_caches : request_caches_) {
+    conn_data_->connection().write(request_caches.first, request_caches.second);
+  }
+}
+
+void UpstreamHandlerImpl::onData(Buffer::Instance& data, bool end_stream) {
+  if (conn_data_) {
+    ENVOY_CONN_LOG(debug, "UpstreamHandlerImpl[{}] data length:{}, end_stream:{}",
+                   conn_data_->connection(), key_, data.length(), end_stream);
+    conn_data_->connection().write(data, end_stream);
+  } else if (upstream_handle_) {
+    ENVOY_LOG(debug, "UpstreamHandler[{}] data length:{}, end_stream:{}, cache msg", key_,
+              data.length(), end_stream);
+    request_caches_.push_back(std::make_pair(Buffer::OwnedImpl(data), end_stream));
```

Comment:
I can see potential problems here:
* Cache could grow big if the connection can't be ready for some time due to any reason
* Concurrent access to the cache might cause conflicts

Can we return " FilterStatus::PauseIteration" here to block incoming requests if connection is not ready?


```c++
ENVOY_LOG(trace, "proxying {} bytes", upstream_request_buffer_.length());
  auto codec = parent_.createCodec();
  codec->encode(*metadata_, *mutation_, upstream_request_buffer_);
  upstream_handler_->onData(upstream_request_buffer_, false);

  if (!upstream_handler_->poolReady()) {
    // Pause while we wait for a connection.
    return FilterStatus::PauseIteration;
  }

  return FilterStatus::ContinueIteration;
  ```